src/dmd/backend/ph2.d

```diff
@@ -0,0 +1,138 @@
+/**
+ * Compiler implementation of the
+ * $(LINK2 http://www.dlang.org, D programming language).
+ *
+ * Copyright:   Copyright (C) 1999-2018 by The D Language Foundation, All Rights Reserved
+ * Authors:     $(LINK2 http://www.digitalmars.com, Walter Bright)
+ * License:     $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Source:      $(LINK2 https://github.com/dlang/dmd/blob/master/src/dmd/backend/ph2.d, backend/ph2.d)
+ */
+
+/* This is only for dmd, not dmc.
+ * It implements a heap allocator that never frees.
+ */
+
+module dmd.backend.ph2;
+
+import core.stdc.stdio;
+import core.stdc.stdlib;
+import core.stdc.string;
+
+import dmd.backend.cc;
+import dmd.backend.global;
+
+extern (C++):
+
+/**********************************************
+ * Do our own storage allocator, a replacement
+ * for malloc/free.
+ */
+
+struct Heap
+{
+    Heap *prev;         // previous heap
+    ubyte *buf;         // buffer
+    ubyte *p;           // high water mark
+    uint nleft;         // number of bytes left
+}
+
+__gshared Heap *heap=null;
+
+void ph_init()
+{
+    if (!heap) {
+        heap = cast(Heap *)calloc(1,Heap.sizeof);
+    }
+    assert(heap);
+}
+
+
+
+void ph_term()
+{
+    //printf("ph_term()\n");
+debug
+{
+    Heap *h;
+    Heap *hprev;
+
+    for (h = heap; h; h = hprev)
+    {
+        hprev = h.prev;
+        free(h.buf);
+        free(h);
+    }
+}
+}
+
+void ph_newheap(size_t nbytes)
+{   uint newsize;
```

Comment:
newline

Reply:
Walter usually waits with the style changes to after the file is converted.